### PR TITLE
[XLA] Simplify transposes that are really reshapes

### PR DESCRIPTION
**Pull request by Keno**
A transpose like
```
f32[1,1,64,1] = transpose(f32[1,64,1,1]), dimensions={3,2,1,0}
```
is really just a reshape (because there's only one non-1 dimension).
Teach algebraic simplifier to make that substitution, to enable applying
reshape-combining optimizations to such instructions.

### CODE REVIEW

**Comment by Keno** [comment]
This change gives about a 15% performance improvement on TPUs when running ResNet50 as implemented in https://github.com/FluxML/MetalHead.jl.

**Comment by Keno** [comment]
Addressed review comment from @davidel.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc [hunk 1]
```diff
@@ -1875,10 +1875,13 @@ Status AlgebraicSimplifierVisitor::HandleReshape(HloInstruction* reshape) {
 
   // Merge reshapes.
   if (HloOpcode::kReshape == operand->opcode()) {
-    return ReplaceWithNewInstruction(
-        reshape, HloInstruction::CreateReshape(reshape->shape(),
-                                               operand->mutable_operand(0)));
+    HloInstruction* new_reshape =
```
Comment (blakehechtman) [inline tensorflow/compiler/xla/service/algebraic_simplifier.cc]: why this change?

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc [hunk 2]
```diff
@@ -2520,6 +2523,13 @@ Status AlgebraicSimplifierVisitor::HandleTranspose(HloInstruction* transpose) {
                                            transpose->dimensions())));
   }
 
+  if (ShapeUtil::TrueRank(transpose->shape()) <= 1) {
+    HloInstruction* reshape = computation_->AddInstruction(
+        HloInstruction::CreateReshape(transpose->shape(), operand));
+    TF_RETURN_IF_ERROR(ReplaceInstruction(transpose, reshape));
+    return HandleReshape(reshape);
```
Comment (blakehechtman) [inline tensorflow/compiler/xla/service/algebraic_simplifier.cc]: algebraic simplifier runs in a fixed point pass, there is not a reason to recurse here?

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc [hunk 3]
```diff
@@ -1770,6 +1770,28 @@ TEST_F(AlgebraicSimplifierTest, TransposesMerged) {
...
+      ROOT reshaped_again = f32[10] reshape(f32[10,1,1] transposed)
+    }
+  )";
+  TF_ASSERT_OK_AND_ASSIGN(
+      auto module,
+      HloRunner::CreateModuleFromString(hlo_string, GetDebugOptionsForTest()));
+
+  AlgebraicSimplifier simplifier(/*is_layout_sensitive=*/false,
+                                 bitcasting_callback());
+  EXPECT_TRUE(simplifier.Run(module.get()).ValueOrDie());
```
Comment (blakehechtman) [inline tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: If you get rid of the recursion yo and execute this many times.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc [hunk 4]
```diff
@@ -2520,6 +2523,13 @@ Status AlgebraicSimplifierVisitor::HandleTranspose(HloInstruction* transpose) {
                                            transpose->dimensions())));
   }
 
+  if (ShapeUtil::TrueRank(transpose->shape()) <= 1) {
```
Comment (blakehechtman) [inline tensorflow/compiler/xla/service/algebraic_simplifier.cc]: Can you generalize this to Detect degenerate dimension, strip them from the transpose and then make remove ones reshape, transpose, add ones reshape. At that point the tranpose becomes a no-op and the reshapes are free to merge in the next fixed point through the algebraic simplifier.

**Comment by Keno** [comment]
Updated to remove the recursion and rely on fix pointing instead. I'll take a look at the more general transform.

**Comment by Keno** [comment]
Added the more general thing as a separate commit.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc [hunk 5]
```diff
@@ -2748,6 +2749,27 @@ Status AlgebraicSimplifierVisitor::HandleSort(HloInstruction* sort) {
   return Status::OK();
 }
 
+static std::vector<int64> DropDegeneratePermDims(const Shape& shape,
+                                                 absl::Span<const int64> perm) {
+  std::vector<int64> dimension_map;
+  int64 i = 0;
+  for (int64 sz : shape.dimensions()) {
+    if (sz == 1) {
+      dimension_map.push_back(-1);
```
Comment (blakehechtman) [inline tensorflow/compiler/xla/service/algebraic_simplifier.cc]: spell out dim_size

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc [hunk 6]
```diff
@@ -2047,6 +2047,27 @@ TEST_F(AlgebraicSimplifierTest, TransposesMerged) {
             computation->root_instruction()->dimensions());
 }
 
+TEST_F(AlgebraicSimplifierTest, TransposeIsReshape) {
+  const char* hlo_string = R"(
+    HloModule module
+
+    ENTRY test {
+      param = f32[10] parameter(0)
```
Comment (blakehechtman) [inline tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: add a test with more than one non-trivial dimension

**Review by blakehechtman — DISMISSED** [review]

**Comment by Keno** [comment]
Before this gets merged, I want to add the test that Blake requested and fix the typo. Will have that shortly, just one sec.

**Comment by Keno** [comment]
Added (and fixed a small bug that Blake pointed out to me in person yesterday).

**Comment by Keno** [comment]
Thanks!
